processor/memorylimiterprocessor/README.md

```diff
@@ -13,21 +13,32 @@ on the configured processors, it is important to put checks in place regarding
 memory usage.
  
 The memory_limiter processor allows to perform periodic checks of memory
-usage if it exceeds defined limits will begin dropping data and forcing GC to reduce
+usage if it exceeds defined limits will begin refusing data and forcing GC to reduce
 memory consumption.
 
 The memory_limiter uses soft and hard memory limits. Hard limit is always above or equal
 the soft limit.
 
-When the memory usage exceeds the soft limit the processor will start dropping the data and
-return errors to the preceding component it in the pipeline (which should be normally a
-receiver).
+When the memory usage exceeds the soft limit the processor will enter the memory limited
+mode and will start refusing the data by returning errors to the preceding component
+in the pipeline that made the ConsumeLogs/Trace/Metrics function call.
+The preceding component should be normally a receiver.
 
-When the memory usage is above the hard limit in addition to dropping the data the
+In memory limited mode the error returned by ConsumeLogs/Trace/Metrics function is a
+non-permanent error. When receivers see this error they are expected to retry sending
+the same data. The receivers may also apply a backpressure to their data sources
+in order to slow down the inflow of data into the Collector and allow the memory usage
+to go below the limits.
+
+>Warning: if the component preceding the memory limiter in the pipeline does not correctly
+retry and send the data again after ConsumeLogs/Trace/Metrics functions return then that
+data will be permanently lost. We consider such components incorrectly implemented.
```

Comment:
Do you mean that receivers need to keep the data for retries? Or applying back-pressure by push-based receivers and responding with 429 applicable here as well? We don't have any receiver that can be called as properly retrying at the moment. I'm not sure we should call them incorrectly implemented until we provide best practices applied in the OTLP receiver.

Reply:
>Do you mean that receivers need to keep the data for retries? Or applying back-pressure by push-based receivers and responding with 429 applicable here as well

Both. It depends on the receiver. For example filelog receiver can simply keep and re-send the last plog.Log, while otlp receiver can drop the received request and simply signal to its client that the client must retry.

It is the receiver's responsibility to ensure the data is retried. How exactly the retrying happens, the mechanics of it are not memory limiter's concern.

>We don't have any receiver that can be called as properly retrying at the moment. I'm not sure we should call them incorrectly implemented until we provide best practices and apply them in the OTLP receiver.

Yes, the goal of this PR is to define best practices :-)

The otlp receiver is [already implemented incorrectly](https://github.com/open-telemetry/opentelemetry-collector/issues/669) regardless of this PR. It does not return the correct status codes to the client, the throttling logic is not implemented correctly.

I can delete the last sentence but it won't change the fact that (as you rightfully call out) most receivers are not correctly implemented :-)

Reply:
Yes, I agree. That documentation should be updated if we keep this warning

Reply:
I will submit a separate PR.